### PR TITLE
#1545: changing DynCollection to LFRicCollection

### LINKED ISSUE #1545: [LFRic] Rename DynCollection and move to psyclone.domain.lfric
In doing #1462 I wanted to put LFRicLoopBounds into psyclone.domain.lfric but, since it needs to import DynCollection from dynamo0p3 this caused circular dependencies.

We need to rename DynCollection -> LFRicCollection and move it into the domain/lfric directory.

### PR DESCRIPTION
**Pull request by mo-lottieturner**
closes #1545

### CODE REVIEW

**Comment by TeranIvy** [comment]
Note: To be merged after #2095 is merged.

**Review by TeranIvy — CHANGES_REQUESTED** [review]
Nice work! There are some things to tidy up, mainly cosmetic (see inline). Please also refresh the branch with the master after the review comments are addressed.

Tests with compilation pass, LFRic examples as well, and developer guide builds and reads well.

**Comment by TeranIvy** [comment]
Back to @mo-lottieturner 😄

File: src/psyclone/domain/lfric/lfric_collection.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors R. W. Ford, A. R. Porter and S. Siso, STFC Daresbury Lab
+# Modified I. Kavcic, A. Coughtrie, L. Turner Met Office
+# Modified J. Henrichs, Bureau of Meteorology
+# Modified A. B. G. Chalk and N. Nobre, STFC Daresbury Lab
+
+''' This module implements the PSyclone Dynamo 0.3 API by 1)
```
[the remark this answers is not in the document]
Comment (mo-lottieturner) [reply src/psyclone/domain/lfric/lfric_collection.py:39]: this was copied across from the dynamo0p3 file - have edited to make it LFRicCollection-specific

File: src/psyclone/domain/lfric/lfric_collection.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+    :raises InternalError: if the supplied node is not a LFRicInvoke or a \
+                           DynKern.
+    '''
+    def __init__(self, node):
+        # Import here to avoid circular dependency
+        # pylint: disable=import-outside-toplevel
+        from psyclone.dynamo0p3 import DynKern
+        if isinstance(node, LFRicInvoke):
+            # We are handling declarations/initialisations for an Invoke
```
[the remark this answers is not in the document]
Comment (mo-lottieturner) [reply src/psyclone/domain/lfric/lfric_collection.py:70]: I thought we were doing no full stops on single line comments?

[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/domain/lfric/lfric_collection.py:70]: You're right! Please ignore this comment. If you see full stops at the end of one-line comments please remove them :)

File: src/psyclone/domain/lfric/lfric_collection.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/domain/lfric/lfric_collection.py:39]: Yes, that's why I suggested "LFRic (Dynamo 0.3) API", so it has LFRic in. We'll remove mentions of Dynamo 0.3 eventually.

**Comment by mo-lottieturner** [comment]
Back to @TeranIvy

File: src/psyclone/domain/lfric/__init__.py [hunk 3]
```diff
@@ -72,6 +73,7 @@
     'KernelInterface',
     'KernStubArgList',
     'LFRicArgDescriptor',
+    'LFRicCollection'
```
Comment (TeranIvy) [inline src/psyclone/domain/lfric/__init__.py:76]: `pylint` says: `__init__.py:76:4: E0603: Undefined variable name 'LFRicCollectionLFRicConstants' in __all__ (undefined-all-variable)`. 

I think there should be a comma :smile:

**Review by TeranIvy — CHANGES_REQUESTED** [review]
Thanks for making changes! There's just one thing to fix (see inline) and the branch needs an update with the master and then it can go on :smile:

**Review by TeranIvy — APPROVED** [review]
All well now. Retested and re-built documentation. Updated changelog and user guide. Proceeding to merge.

**Pull request merged**